bazel/testing/lit_autoupdate_base.py

```diff
@@ -0,0 +1,408 @@
+        else:
+            check_line = check_lines.pop(0)
+            # Indent to match the next original line.
+            check_line.indent = re.findall("^ *", orig_lines[0].text)[0]
+            result_lines.append(check_line)
+    # One list is non-empty; append remaining lines from both to catch it.
+    result_lines.extend(orig_lines)
+    result_lines.extend(check_lines)
+
+    return result_lines
+
+
+def update_check(update_args: UpdateArgs, test: str) -> bool:
+    """Updates the CHECK: lines for `test` by running explorer.
+
+    Returns true if a change was made.
+    """
+    with open(test) as f:
+        orig_lines = f.readlines()
+
+    # Make sure we're supposed to autoupdate.
+    autoupdate = find_autoupdate(test, orig_lines)
+    if autoupdate is None:
+        return False
+
+    # Determine the merged output lines.
+    out_lines = get_actual_output(
+        update_args, test, autoupdate.cmd, update_args.extra_check_replacements
+    )
+    result_lines = merge_lines(
+        update_args.line_number_format,
+        update_args.line_number_pattern,
+        autoupdate.line_number,
+        orig_lines,
+        out_lines,
+    )
+
+    # Calculate the remap for original lines.
+    line_number_remap = dict(
+        [
+            (line.line_number, i)
+            for i, line in enumerate(result_lines)
+            if isinstance(line, OriginalLine)
+        ]
+    )
+    # If the last line of the original output was a CHECK, replace it with an
+    # empty line.
+    if orig_lines[-1].lstrip().startswith("// CHECK"):
+        line_number_remap[len(orig_lines) - 1] = len(result_lines) - 1
+
+    # Generate contents for any lines that depend on line numbers.
+    formatted_result_lines = [
+        line.format(output_line_number=i, line_number_remap=line_number_remap)
+        for i, line in enumerate(result_lines)
+    ]
+
+    # If nothing's changed, we're done.
+    if formatted_result_lines == orig_lines:
+        return False
+
+    # Interleave the new CHECK: lines with the tested content.
+    with open(test, "w") as f:
+        f.writelines(formatted_result_lines)
+        return True
+
+
+def update_checks(update_args: UpdateArgs, tests: Set[str]) -> None:
+    """Runs bazel to update CHECK: lines in lit tests."""
+
+    def map_helper(test: str) -> bool:
+        updated = update_check(update_args, test)
+        print(".", end="", flush=True)
+        return updated
+
+    print(f"Updating {len(tests)} lit test(s)...")
+    with futures.ThreadPoolExecutor() as exec:
+        # list() iterates in order to immediately propagate exceptions.
+        results = list(exec.map(map_helper, tests))
+
+    # Each update call indicates progress with a dot without a newline, so put a
+    # newline to wrap.
+    print(f"\nUpdated {results.count(True)} lit test(s).")
+
+
+def main() -> None:
+    # Go to the repository root so that paths will match bazel's view.
+    os.chdir(Path(__file__).parent.parent.parent)
```

Comment:
Presumably this means that the test name arguments are interpreted relative to the repository root, which seems undesirable. Should we parse the args and `realpath` or `abspath` the `parsed_args.tests` list first?

Reply:
Done, with some shifts in logic to use pathlib.Path more.

Reply:
(I felt it'd be clearer if parse_args() encapsulated more)